
### Supervised Learning
### Activity: Building a Student Intervention System

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail or pass. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: ** 

Classification problem. Our job is to find out whether the student needs intevention or not

In Regression, we try to find the best fit line, which can predict the output more accurately.but in Classification we try to find the decision boundary which can divide the dataset into different classes(discrete type of output).Here our goal for this project is to identify students who might need early intervention before they fail or pass,and here we cant predict our output more accurately.so the type of supervised learning what we follow is classification supervised algorithm.

### Question-2
load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [1]:
# Import libraries
import numpy as np
import pandas as pd
import warnings 
warnings.filterwarnings("ignore")

In [2]:
# Read student data
data=pd.read_csv('student-data.csv')

In [3]:
data.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  internet romantic  famrel  freetime  goout Dalc Walc health absences passed  
0       no       no       4         3      4    1    1      3        6     no  
1      yes       no       5         3      3    1    1      3        4     no  
2      yes       no       4         3      2    2    3      3       10    yes  
3      yes      yes       3         2      2    1    1      5        2    yes  
4       no       no       4         3      2    1    2      5        4    yes  

[5 rows x 31 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 31 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

In [5]:
# dropping duplicate rows if any 

data=data.drop_duplicates()
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 395 entries, 0 to 394
Data columns (total 31 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

### Question-3
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [6]:
# Calculate number of students
n_students=data.shape[0]
n_students

395

In [7]:
# Calculate number of features
n_features=len(data.columns)
n_features

31

In [8]:
# Calculate passing students
n_passed=data[data['passed']=='yes'].shape[0]
n_passed

265

In [9]:
# Calculate failing students
n_failed=data['passed'].value_counts().no
n_failed

130

In [10]:
# Calculate graduation rate

grad_rate=( n_passed/n_students)*100
grad_rate

67.08860759493672

Graduation rate is the number of students graduating divided by the number of students. ie n_passed/n_students. To convert into percentage, multiply by 100

In [11]:
# Print the results
print('The total number of students:', n_students)
print('The total number of features for each student: ', n_features)
print('The number of those students who passed: ', n_passed)
print('The number of those students who failed: ', n_failed)
print('The graduation rate of the class in percent (%):',round(grad_rate,2))

The total number of students: 395
The total number of features for each student:  31
The number of those students who passed:  265
The number of those students who failed:  130
The graduation rate of the class in percent (%): 67.09


## Preparing the Data
you will prepare the data for modeling, training and testing.

### Question-4 Identify feature and target columns


separate the student data into feature and target columns to see if any features are non-numeric.

In [12]:
# Extract feature columns

In [13]:
data.loc[:, data.columns != 'passed']

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0    ...    yes       no        no       4         3     4    1    1      3   
1    ...    yes      yes        no       5         3     3    1    1      3   
2    ...    yes      yes        no       4         3     2    2    3      3   
3    ...    yes      yes       yes       3         2     2    1    1      5   
4    ...    yes       no        no       4         3     2    1    2      5   
..   ...    ...      ...       ...     ...       ...   ...  ...  ...    ...   
390  ...    yes       no        no       5         5     4    4    5      4   
391  ...    yes      yes        no       2         4     5    3    4      2   
392  ...    yes       no        no       5         5     3    3    3      3   
393  ...    yes      yes        no       4         4     1    3    4      5   
394  ...    yes      yes        no       3         2     3    3    3      5   

    absences  
0          6  
1          4  
2         10  
3          2  
4          4  
..       ...  
390       11  
391        3  
392        3  
393        0  
394        5  

[395 rows x 30 columns]

In [14]:
# Extract target column 'passed'

In [15]:
pd.DataFrame(data['passed'])

passed
0       no
1       no
2      yes
3      yes
4      yes
..     ...
390     no
391    yes
392     no
393    yes
394     no

[395 rows x 1 columns]

In [16]:
# Separate the data into feature data and target data (X and y, respectively)

In [17]:
x=data.drop(['passed'],axis=1)
y=data['passed']

### Question-5 Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [18]:
for (columnName, columnData) in data.iteritems():
      print(columnData.value_counts())

GP    349
MS     46
Name: school, dtype: int64
F    208
M    187
Name: sex, dtype: int64
16    104
17     98
18     82
15     82
19     24
20      3
22      1
21      1
Name: age, dtype: int64
U    307
R     88
Name: address, dtype: int64
GT3    281
LE3    114
Name: famsize, dtype: int64
T    354
A     41
Name: Pstatus, dtype: int64
4    131
2    103
3     99
1     59
0      3
Name: Medu, dtype: int64
2    115
3    100
4     96
1     82
0      2
Name: Fedu, dtype: int64
other       141
services    103
at_home      59
teacher      58
health       34
Name: Mjob, dtype: int64
other       217
services    111
teacher      29
at_home      20
health       18
Name: Fjob, dtype: int64
course        145
home          109
reputation    105
other          36
Name: reason, dtype: int64
mother    273
father     90
other      32
Name: guardian, dtype: int64
1    257
2    107
3     23
4      8
Name: traveltime, dtype: int64
2    198
1    105
3     65
4     27
Name: studytime, dtype: int64
0    312
1  

In [19]:
for i in ['schoolsup','famsup','paid','activities','nursery','higher','internet','romantic','passed']:
    data[i]=data[i].map({'yes':1 ,'no':0})

In [20]:
for (columnName, columnData) in data.iteritems():
      print(columnData.value_counts())

GP    349
MS     46
Name: school, dtype: int64
F    208
M    187
Name: sex, dtype: int64
16    104
17     98
18     82
15     82
19     24
20      3
22      1
21      1
Name: age, dtype: int64
U    307
R     88
Name: address, dtype: int64
GT3    281
LE3    114
Name: famsize, dtype: int64
T    354
A     41
Name: Pstatus, dtype: int64
4    131
2    103
3     99
1     59
0      3
Name: Medu, dtype: int64
2    115
3    100
4     96
1     82
0      2
Name: Fedu, dtype: int64
other       141
services    103
at_home      59
teacher      58
health       34
Name: Mjob, dtype: int64
other       217
services    111
teacher      29
at_home      20
health       18
Name: Fjob, dtype: int64
course        145
home          109
reputation    105
other          36
Name: reason, dtype: int64
mother    273
father     90
other      32
Name: guardian, dtype: int64
1    257
2    107
3     23
4      8
Name: traveltime, dtype: int64
2    198
1    105
3     65
4     27
Name: studytime, dtype: int64
0    312
1  

Encoding all categorical features using one hot encoding

In [21]:
data.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'passed'],
      dtype='object')

In [22]:
for i in ['school', 'sex', 'address', 'famsize', 'Pstatus','Mjob', 'Fjob', 'reason', 'guardian']:
    x=pd.get_dummies(data[i])
    data=pd.concat([data,x],axis=1)

In [23]:
data

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ...  other services  teacher  course  home  other  reputation  father  \
0    ...      0        0        1       1     0      0           0       0   
1    ...      1        0        0       1     0      0           0       1   
2    ...      1        0        0       0     0      1           0       0   
3    ...      0        1        0       0     1      0           0       0   
4    ...      1        0        0       0     1      0           0       1   
..   ...    ...      ...      ...     ...   ...    ...         ...     ...   
390  ...      0        1        0       1     0      0           0       0   
391  ...      0        1        0       1     0      0           0       0   
392  ...      1        0        0       1     0      0           0       0   
393  ...      1        0        0       1     0      0           0       0   
394  ...      0        0        0       1     0      0           0       1   

     mother  other  
0         1      0  
1         0      0  
2         1      0  
3         1      0  
4         0      0  
..      ...    ...  
390       0      1  
391       1      0  
392       0      1  
393       1      0  
394       0      0  

[395 rows x 58 columns]

In [24]:
data.shape

(395, 58)

In [25]:
data.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'passed', 'GP', 'MS', 'F', 'M', 'R', 'U',
       'GT3', 'LE3', 'A', 'T', 'at_home', 'health', 'other', 'services',
       'teacher', 'at_home', 'health', 'other', 'services', 'teacher',
       'course', 'home', 'other', 'reputation', 'father', 'mother', 'other'],
      dtype='object')

In [26]:
data=data.drop(['school', 'sex', 'address', 'famsize', 'Pstatus','Mjob', 'Fjob', 'reason', 'guardian'],axis=1)

In [27]:
data.head()

age  Medu  Fedu  traveltime  studytime  failures  schoolsup  famsup  paid  \
0   18     4     4           2          2         0          1       0     0   
1   17     1     1           1          2         0          0       1     0   
2   15     1     1           1          2         3          1       0     1   
3   15     4     2           1          3         0          0       1     1   
4   16     3     3           1          2         0          0       1     1   

   activities  ...  other  services  teacher  course  home  other  reputation  \
0           0  ...      0         0        1       1     0      0           0   
1           0  ...      1         0        0       1     0      0           0   
2           0  ...      1         0        0       0     0      1           0   
3           1  ...      0         1        0       0     1      0           0   
4           0  ...      1         0        0       0     1      0           0   

   father  mother  other  
0       0       1      0  
1       1       0      0  
2       0       1      0  
3       0       1      0  
4       1       0      0  

[5 rows x 49 columns]

In [28]:
data.columns

Index(['age', 'Medu', 'Fedu', 'traveltime', 'studytime', 'failures',
       'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher',
       'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc',
       'health', 'absences', 'passed', 'GP', 'MS', 'F', 'M', 'R', 'U', 'GT3',
       'LE3', 'A', 'T', 'at_home', 'health', 'other', 'services', 'teacher',
       'at_home', 'health', 'other', 'services', 'teacher', 'course', 'home',
       'other', 'reputation', 'father', 'mother', 'other'],
      dtype='object')

In [29]:
data.shape

(395, 49)

In [30]:
data.dtypes

age           int64
Medu          int64
Fedu          int64
traveltime    int64
studytime     int64
failures      int64
schoolsup     int64
famsup        int64
paid          int64
activities    int64
nursery       int64
higher        int64
internet      int64
romantic      int64
famrel        int64
freetime      int64
goout         int64
Dalc          int64
Walc          int64
health        int64
absences      int64
passed        int64
GP            uint8
MS            uint8
F             uint8
M             uint8
R             uint8
U             uint8
GT3           uint8
LE3           uint8
A             uint8
T             uint8
at_home       uint8
health        uint8
other         uint8
services      uint8
teacher       uint8
at_home       uint8
health        uint8
other         uint8
services      uint8
teacher       uint8
course        uint8
home          uint8
other         uint8
reputation    uint8
father        uint8
mother        uint8
other         uint8
dtype: object

Now the new dataset contain only numeric values with 49 columns and  395 entries

In [31]:
# Updating x and y again

x=data.drop(['passed'],axis=1)
y=data['passed']

### Question - 6 Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. you will need to implement the following:
- Randomly shuffle and split the data (`X`, `y`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [32]:
# splitting the data into train and test

In [33]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(x,y,random_state=42,test_size=0.24)

In [34]:
# Show the results of the split

In [35]:
X_train.shape

(300, 48)

In [36]:
y_train.shape

(300,)

In [37]:
X_test.shape

(95, 48)

In [38]:
y_test.shape

(95,)

### Question - 7  Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data and measure the accuracy score.

###  Model Application
*List three supervised learning models that are appropriate for this problem. What are the general applications of each model? What are their strengths and weaknesses? Given what you know about the data, why did you choose these models to be applied?*

Popular algorithms that can be used for binary classification include:

Logistic Regression
k-Nearest Neighbors
Decision Trees
Support Vector Machine

1. Logistic Regression
        Pros:
        1. Good performance with small dataset 
        2. Its output can be interpreted as a probabilit
        Cons:
        1. It can only provide linear solutions
        
2. Support Vector Machine:
       Pros:
       It can provide non-linear solutions      
       Cons:
       to achieve good performance, they require knowledge about the  kernel employed
       
3. Decision Trees:
        Pros:
        They can handle categorixcal features
        They perform wellin datasets with large number of features
        Cons:
        Interpretability is questionable

In [39]:
# Import the three supervised learning models from sklearn

from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier

from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

In [40]:
# fit model-1  on traning data 

In [41]:
lr = LogisticRegression()
lr.fit(X_train, y_train)

LogisticRegression()

In [42]:
# predict on the test data 

In [43]:
y_pred = lr.predict(X_test)

In [44]:
# calculate the accuracy score

In [45]:
print('accuracy is',accuracy_score(y_pred,y_test))

accuracy is 0.7157894736842105


In [46]:
# fit the model-2 on traning data and predict on the test data and measure the accuracy

In [47]:
svm = SVC()
svm.fit(X_train, y_train)

y_pred = svm.predict(X_test)


# Accuracy score

print('accuracy is',accuracy_score(y_pred,y_test))

accuracy is 0.631578947368421


In [48]:
# fit the model-3 on traning data and predict on the test data and measure the accuracy

In [49]:
dt = DecisionTreeClassifier()

dt.fit(X_train, y_train)

y_pred = dt.predict(X_test)


# Accuracy score

print('accuracy is',accuracy_score(y_pred,y_test))

accuracy is 0.5263157894736842


#### Inference:
    Logistic Regression have better accuracy in prediction than Support Vector Machine and Decision Tree